L1Trigger/DemonstratorTools/plugins/GTTFileWriter.cc

```diff
@@ -132,36 +176,107 @@ GTTFileWriter::GTTFileWriter(const edm::ParameterSet& iConfig)
                                  kMaxLinesPerFile,
                                  kChannelIdsInput,
                                  kChannelSpecsInput),
+      fileWriterSelectedTracks_(l1t::demo::parseFileFormat(iConfig.getUntrackedParameter<std::string>("format")),
+                                iConfig.getUntrackedParameter<std::string>("selectedTracksFilename"),
+                                kFramesPerTMUXPeriod,
+                                kGTTBoardTMUX,
+                                kMaxLinesPerFile,
+                                kChannelIdsInput,
+                                kChannelSpecsInput),
+      fileWriterVertexAssociatedTracks_(
+          l1t::demo::parseFileFormat(iConfig.getUntrackedParameter<std::string>("format")),
+          iConfig.getUntrackedParameter<std::string>("vertexAssociatedTracksFilename"),
+          kFramesPerTMUXPeriod,
+          kGTTBoardTMUX,
+          kMaxLinesPerFile,
+          kChannelIdsInput,
+          kChannelSpecsInput),
       fileWriterOutputToCorrelator_(l1t::demo::parseFileFormat(iConfig.getUntrackedParameter<std::string>("format")),
-                                    iConfig.getUntrackedParameter<std::string>("outputFilename"),
+                                    iConfig.getUntrackedParameter<std::string>("outputCorrelatorFilename"),
                                     kFramesPerTMUXPeriod,
                                     kGTTBoardTMUX,
                                     kMaxLinesPerFile,
-                                    kChannelSpecsOutputToCorrelator) {}
+                                    kChannelSpecsOutputToCorrelator),
+      fileWriterOutputToGlobalTrigger_(l1t::demo::parseFileFormat(iConfig.getUntrackedParameter<std::string>("format")),
+                                       iConfig.getUntrackedParameter<std::string>("outputGlobalTriggerFilename"),
+                                       kFramesPerTMUXPeriod,
+                                       kGTTBoardTMUX,
+                                       kMaxLinesPerFile,
+                                       kChannelIdsOutputToGlobalTrigger,
+                                       kChannelSpecsOutputToGlobalTrigger) {}
 
 void GTTFileWriter::analyze(const edm::Event& iEvent, const edm::EventSetup& iSetup) {
   using namespace edm;
   using namespace l1t::demo::codecs;
 
-  // 1) Encode track information onto vectors containing link data
-  const auto trackData(encodeTracks(iEvent.get(tracksToken_)));
-  const auto convertedTrackData(encodeTracks(iEvent.get(convertedTracksToken_)));
-  const auto outputData(encodeVertices(iEvent.get(verticesToken_)));
+  // 0) Gather the necessary collections
+  const auto& tracksCollection = iEvent.get(tracksToken_);
+  const auto& convertedTracksCollection = iEvent.get(convertedTracksToken_);
```

Comment:
It is curious that a `convertedTracksToken_` is used to get a `convertedTracksCollection` , while a `convertedTrackCollectionToken_` is used to get a `convertedTracksHandle` in [line 223](https://github.com/cms-sw/cmssw/pull/39422/files#diff-d5902b75114772d43c8686fe8029e6856e52201154e0b2e7d6682ee96e872c2bR223)

Could you please confirm that this is really what intended?

Reply:
@aperloff Could you please answer this question?

Reply:
There was a need to have both an `edm::Handle` to the "convertedTracksCollection" (L223) and the collection itself (L214). While the logical choice is to have both of these collections come from the same `edm::EDGetTokenT` (L153 and L155), this is not a technical requirement; the `edm::Handle` could come from one token and the collection from another, as is done here. I chose to go with maximal flexibility, but put in a logical framework by setting the two tokens to point to the same option from the configuration file.

Reply:
> There was a need to have both an `edm::Handle` to the "convertedTracksCollection" (L223) and the collection itself (L214). While the logical choice is to have both of these collections come from the same `edm::EDGetTokenT` (L153 and L155), this is not a technical requirement; the `edm::Handle` could come from one token and the collection from another, as is done here. I chose to go with maximal flexibility, but put in a logical framework by setting the two tokens to point to the same option from the configuration file.

I find the crossed names quite confusing... But ok, at least you checked that it was not done as such by mistake.
Please go ahead with the other fixes